AI #1793: Create Commitment Program Eligibility Details column (#1882)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/commitmentprogrameligibilitydetails.md

Reviewer: Have a look at the build errors [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/actions/runs/23444012928/job/68202557624?pr=1882).

There are quite a few of them, but they're all straightforward and minor.  Many, but not all, of them are related to the point @ijurica made about using dashes instead of asterisks.  But there are a few others to address as well.

You'll see the errors reference lines in `spec.md`, which is the final output of the spec.  You'll have to fetch `spec.md` to see the relevant lines.  You can either get it from the GitHub Action [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/actions/runs/23444012928?pr=1882), or you can get it locally if you're doing this work on your local machine.

@Matt-Cowsert can walk you through how to address these.

Including a small snippet below so you can get a sense of what these look like:

```
spec.md:3881:5: MD030: Spaces after list markers [Expected: 1; Actual: 3] (list-marker-space)
spec.md:3882:5: MD007: Unordered list indentation [Expected: 2, Actual=4] (ul-indent)
spec.md:3882:5: MD030: Spaces after list markers [Expected: 1; Actual: 3] (list-marker-space)
spec.md:3883:5: MD007: Unordered list indentation [Expected: 2, Actual=4] (ul-indent)
spec.md:3883:5: MD030: Spaces after list markers [Expected: 1; Actual: 3] (list-marker-space)
spec.md:3884:5: MD007: Unordered list indentation [Expected: 2, Actual=4] (ul-indent)
spec.md:3884:5: MD030: Spaces after list markers [Expected: 1; Actual: 3] (list-marker-space)
spec.md:3952:1: MD036: Emphasis possibly used instead of a heading element. (no-emphasis-as-heading,no-emphasis-as-header)
spec.md:3960:1: MD036: Emphasis possibly used instead of a heading element. (no-emphasis-as-heading,no-emphasis-as-header)
spec.md:3968:1: MD036: Emphasis possibly used instead of a heading element. (no-emphasis-as-heading,no-emphasis-as-header)
```

Author: These have been addressed.

Reviewer: Reposting [this outdated thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1882/files/aa5cf6a98e84ad797bf90c7318947695ade6853e#r2937132479) for visibility.

@SanjnaSrivatsaVMware
@Matt-Cowsert has added a calculation for Coverage in the supported feature for this column PR https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2025.

We should definitely add one or more concrete examples here, demonstrating how that calculation would work with hard numbers.  Could you partner with @Matt-Cowsert to either augment some of the examples here, or add a couple at the end?

Something like this (note: AI-generated, do not just copy/paste).  Hopefully it gets the point across for what to _do_ with this column, not just show how it might be formatted.

----

### Example Coverage Calculations

The primary value of the `CommitmentEligibilityDetails` column is adjusting the denominator of a commitment coverage calculation to exclude ineligible spend. This provides a highly actionable coverage metric.

**Example 1: Compute and Storage (Provider A)**

Consider the following sample usage data for Provider A:

| ServiceProviderName | ServiceName | ChargeCategory | EffectiveCost | CommitmentDiscountId | CommitmentEligibilityDetails |
|:---|:---|:---|:---|:---|:---|
| Provider A | Virtual Machines | Usage | 10.00 | `arn:providera:savingsplans...` | `{"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}]}` |
| Provider A | Virtual Machines | Usage | 15.00 | null | `{"CommitmentPrograms": [{"ProgramType": "SavingsPlan"}]}` |
| Provider A | ObjectStorage | Usage | 25.00 | null | null |

By isolating the denominator to only rows where `CommitmentEligibilityDetails` is not null, the $25.00 of ineligible ObjectStorage spend is correctly excluded:

* **Eligibility-Adjusted Coverage:** 10.00 Covered / 25.00 Total Eligible Spend = **40% Coverage**

**Example 2: Observability Platform (Provider B)**

Consider the following sample usage data for Provider B, where infrastructure hosts are eligible for monthly commitments, but custom metrics are not:

| ServiceProviderName | ServiceName | ChargeCategory | EffectiveCost | CommitmentDiscountId | CommitmentEligibilityDetails |
|:---|:---|:---|:---|:---|:---|
| Provider B | Infrastructure | Usage | 100.00 | `sub_12345` | `{"CommitmentPrograms": [{"ProgramType": "MonthlyCommitment"}]}` |
| Provider B | Infrastructure | Usage | 200.00 | null | `{"CommitmentPrograms": [{"ProgramType": "MonthlyCommitment"}]}` |
| Provider B | Custom Metrics | Usage | 50.00 | null | null |

Excluding the $50.00 of ineligible Custom Metrics spend yields the true coverage rate for eligible services:

* **Eligibility-Adjusted Coverage:** 100.00 Covered / 300.00 Total Eligible Spend = **33.3% Coverage**

Reviewer 2: @shawnalpay I've added the examples down below, one for Commitment Program Eligibility coverage rate and one for uncovered eligible spend by Commitment Program, program type.

Reviewer 2: Resolving as complete

Reviewer: Please update the PR summary in the [parent post of the Conversation tab](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1882#issue-3848576296).  It is currently outdated and therefore misleading (e.g., doesn't include the column rename).

Reviewer 2: The PR description was updated on April 3 with the current column name, design decisions, dependency table, and change log. Let me know if anything specific still needs updating.

Reviewer: Hmm.  You are right.  I must have been looking at a stale browser tab.

Resolving.

## Requirements

CommitmentProgramEligibilityDetails MUST adhere to the following requirements:

* CommitmentProgramEligibilityDetails MUST be of type String.
* CommitmentProgramEligibilityDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* CommitmentProgramEligibilityDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* CommitmentProgramEligibilityDetails MUST NOT be null when a charge is eligible for a [*commitment program*](#glossary:commitment-program), regardless of whether a [*commitment*](#glossary:commitment) was actually applied to the charge.
* CommitmentProgramEligibilityDetails MUST NOT consider transient account configurations or quotas that might temporarily prevent purchase or participation in a *commitment program*.
* CommitmentProgramEligibilityDetails MUST include all publicly available *commitment programs* for which the usage is eligible.
* CommitmentProgramEligibilityDetails MAY include negotiated *commitment programs* when the usage is eligible and the program is not broadly applicable across the service provider's service catalog.

> 

Reviewer: Remove empty line.

```suggestion
```

* CommitmentProgramEligibilityDetails MUST NOT include data related to *commitment* [*periods*](#glossary:period) or payment options.

> * CommitmentProgramEligibilityDetails MUST conform to [CommitmentProgramEligibilityDetailsObject](#datasets.costandusage.CommitmentProgramEligibilityDetails.CommitmentProgramEligibilityDetailsobject) requirements when CommitmentProgramEligibilityDetails is not null.

Reviewer: Anchors are all lowercase.

```suggestion
* CommitmentProgramEligibilityDetails MUST conform to [CommitmentProgramEligibilityDetailsObject](#datasets.costandusage.commitmentprogrameligibilitydetails.commitmentprogrameligibilitydetailsobject) requirements when CommitmentProgramEligibilityDetails is not null.
```

  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST NOT be null.
  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider (e.g., "FlexibleSpendPlan", "ResourceReservation", "BulkCapacityCredit", "AdvanceResourceCommitment").
  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST equal [CommitmentDiscountType](#datasets.costandusage.commitmentdiscounttype) for one object in CommitmentProgramEligibilityDetailsObject.CommitmentPrograms when CommitmentDiscountType is not null.
  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms.ProgramType SHOULD correspond to terminology disclosed by the service provider in public documentation.

### Top-Level Properties

| Property             | Type  | Required | Description                                                                         |
|:----------|:----------|:----------|:---------------------------------------|
| `CommitmentPrograms` | Array | True     | Array of objects identifying *commitment programs* for which the usage is eligible. |

> ### Example Entries

Reviewer: This isn't a list of example entries.  It's a documentation of what goes into the CommitmentPrograms Object.  Further, we should describe what this is (see [AllocatedMethodDetails](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/cost_and_usage/columns/allocatedmethoddetails.md) for an example of how to describe this):

```suggestion
### CommitmentPrograms Object

The `CommitmentPrograms` array contains one or more objects, each of which contains the following entries:
```

## Requirements

CommitmentProgramEligibilityDetails MUST adhere to the following requirements:

* CommitmentProgramEligibilityDetails MUST be of type String.
* CommitmentProgramEligibilityDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* CommitmentProgramEligibilityDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* CommitmentProgramEligibilityDetails MUST NOT be null when a charge is eligible for a [*commitment program*](#glossary:commitment-program), regardless of whether a [*commitment*](#glossary:commitment) was actually applied to the charge.

> * CommitmentProgramEligibilityDetails MUST NOT consider transient account configurations or quotas that might temporarily prevent purchase or participation in a *commitment program*.

Reviewer: Per the `AGENTS.md` guidelines for this repository:
> **State vs Behavior:** Normative requirements MUST describe a verifiable state, not an operational behavior. Do not use process-oriented verbs (e.g., *ensure*, *handle*, *support*, *provide*) as the primary constraint...

The word "consider" acts as an operational/process-oriented verb. We should update this to describe the verifiable state of the data instead. 

Also (and this is non-blocking), I feel like "transient account configurations or quotas" is a touch too specific.  Recommend using the word "restrictions" and moving those out to an `e.g.,`.

**Suggested fix:**
```suggestion
* CommitmentProgramEligibilityDetails MUST NOT reflect restrictions (e.g., transient account configurations, quotas) that might temporarily prevent purchase or participation in a *commitment program*.
```

* CommitmentProgramEligibilityDetailsObject MUST have property keys that begin with the string "x_" unless it is a FOCUS-defined property key.
* CommitmentProgramEligibilityDetailsObject.CommitmentPrograms MUST adhere to the following requirements:
  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms MUST be of type Array.
  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms MUST contain one or more objects.
  * Each entry in CommitmentProgramEligibilityDetailsObject.CommitmentPrograms MUST be of type JSON Object.
  * Each entry in CommitmentProgramEligibilityDetailsObject.CommitmentPrograms MUST have a property key "ProgramType".
  * Each entry in CommitmentProgramEligibilityDetailsObject.CommitmentPrograms MAY contain additional data generator-defined property keys.
  * Each entry in CommitmentProgramEligibilityDetailsObject.CommitmentPrograms MUST have property keys that begin with the string "x_" unless it is a FOCUS-defined property key.
  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST be of type String.
  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST NOT be null.

>   * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider (e.g., "FlexibleSpendPlan", "ResourceReservation", "BulkCapacityCredit", "AdvanceResourceCommitment").

Reviewer: As I suggested in [this outdated thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1882#discussion_r3054809097) reopened today by @flanakin, one path forward here is to simply strike these examples from the normative requirement itself.

That doesn't address the full thrust of @flanakin's concern (also described in [this other open thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1882#discussion_r3054816480)), which we would look to tackle in consistency review if @SanjnaSrivatsaVMware and @Matt-Cowsert are in support.  But I believe this change would at least let us move forward with approval tomorrow, pending @flanakin's agreement, and all other changes would be `Informative` instead of `Normative`.

```suggestion
  * CommitmentProgramEligibilityDetailsObject.CommitmentPrograms.ProgramType MUST correspond to a *commitment program* type supported by the service provider.
```

Author: I understand this is a non-trivial change that we will have to make during the consistency review, but this workaround is good.

## Requirements

CommitmentProgramEligibilityDetails MUST adhere to the following requirements:

* CommitmentProgramEligibilityDetails MUST be of type String.
* CommitmentProgramEligibilityDetails MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* CommitmentProgramEligibilityDetails MUST conform to [JsonObjectFormat](#attributes.jsonobjectformat) requirements.
* CommitmentProgramEligibilityDetails MUST NOT be null when a charge is eligible for a [*commitment program*](#glossary:commitment-program), regardless of whether a [*commitment*](#glossary:commitment) was actually applied to the charge.
* CommitmentProgramEligibilityDetails MUST NOT consider transient account configurations or quotas that might temporarily prevent purchase or participation in a *commitment program*.
* CommitmentProgramEligibilityDetails MUST include all publicly available *commitment programs* for which the usage is eligible.

> * CommitmentProgramEligibilityDetails SHOULD include negotiated *commitment programs* for which the usage is eligible.

Reviewer 2: ## TF3 Action Item - 2026-04-03

The group decided to change requirement 15 from SHOULD to MAY for negotiated commitment programs. The current wording is too broad and could force providers to include globally applicable programs (e.g., OCI Universal Credits) on every row, creating data bloat. Matt committed to providing a GitHub suggestion with tightened language that conveys: "MAY include negotiated commitment programs when [condition]." The exact condition is unresolved but should communicate that broadly applicable negotiated programs don't need to appear on every line item.

**Scope:**
- [x] Draft and post a GitHub suggestion changing "SHOULD include negotiated commitment programs for which the usage is eligible" to a MAY statement with a qualifying condition
- [x] Ensure the phrasing aligns with Riley's framing: providers MAY include negotiated programs, but globally applicable ones can be handled via the commitment dataset instead

**Assignee:** @matt-cowsert

> * CommitmentProgramEligibilityDetails SHOULD include negotiated *commitment programs* for which the usage is eligible.

Reviewer 2: ```suggestion
* CommitmentProgramEligibilityDetails MAY include negotiated *commitment programs* when the usage is eligible and the program is not broadly applicable across the service provider's service catalog.

```

| `CommitmentPrograms` | Array | True     | Array of objects identifying *commitment programs* for which the usage is eligible. |

### Example Entries

| Key         | ValueType                            | Required | Description                                                                                                |
|:-------------|:-------------|:-------------|:------------------------------|
| ProgramType | [String](#attributes.stringhandling) | True     | The specific type of commitment program (e.g., discount or capacity reservation) available for this usage. |

### Object Example

``` json
{
  "CommitmentPrograms": [
    { "ProgramType": "FlexibleSpendPlan" },
    { "ProgramType": "ResourceReservation" },
    { "ProgramType": "AdvanceResourceCommitment" },
    { "ProgramType": "ZonalResourceCommitment" }
  ]
}
```

### JSON Type Definition

``` json
{
  "definitions": {
    "commitmentProgramEntry": {
      "properties": {
        "ProgramType": { "type": "string" }
      }
    }
  },
  "properties": {
    "CommitmentPrograms": {
      "elements": { "ref": "commitmentProgramEntry" }
    }
  }

> }

Reviewer: The normative requirements explicitly allow data generators to include custom properties (prefixed with `x_`) at both the top level and within the `CommitmentPrograms` array entries. However, in the JTD, objects strictly forbid undeclared properties by default. Because the current JTD omits `"additionalProperties": true`, any provider who actually includes an allowed `x_` key will fail schema validation.

Here is the corrected JTD block that allows the custom properties to pass validation:

```suggestion
{
  "definitions": {
    "commitmentProgramEntry": {
      "properties": {
        "ProgramType": { "type": "string" }
      },
      "additionalProperties": true
    }
  },
  "properties": {
    "CommitmentPrograms": {
      "elements": { "ref": "commitmentProgramEntry" }
    }
  },
  "additionalProperties": true
}
```

Author: Makes sense Shawn. Accepting this suggestion

| ServiceProviderName | ServiceName | CommitmentProgramEligibilityDetails                                                                                                                                                                           |
|---------------------|-------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Aura Web            | Compute     | {"CommitmentPrograms": [{"ProgramType": "FlexibleSpendPlan"}, {"ProgramType": "ResourceReservation"}, {"ProgramType": "AdvanceResourceCommitment"}, {"ProgramType": "ZonalResourceCommitment"}]} |

### Coverage Rate with Eligibility-Adjusted Denominator

This example demonstrates how to calculate an accurate [*commitment*](#glossary:commitment) coverage rate using [CommitmentProgramEligibilityDetails](#datasets.costandusage.CommitmentProgramEligibilityDetails) alongside [CommitmentDiscountId](#datasets.costandusage.commitmentdiscountid).

> Acme Corp runs compute workloads on Aura Web. Some usage is covered by a Resource Reservation, some is eligible but uncovered, and a support fee is ineligible for any [*commitment program*](#glossary:commitment-program). While the specification mandates that CommitmentProgramEligibilityDetails must not be null when a charge is eligible, practitioners may encounter non-compliant provider data where this column is omitted on covered rows. This creates a gap in naive coverage calculations that rely solely on CommitmentProgramEligibilityDetails to define the eligible population.

Reviewer: Spec documents usually strictly define the "ideal/compliant" state. Acknowledging that providers will break the rule -- and providing the fix here -- might inadvertently give providers an "out." Should we move this specific non-compliant workaround to a separate practitioner playbook or use-case guide, and keep the spec example focused strictly on compliant data?

Note that this comment conflicts with one of my previous comments on #2025 [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2025/files#r3035976676).  I think we either have to double-down on providing context on potentially non-compliant data, or (perhaps better?) we should strike any such references.

Reviewer 2: I went back and forth with @ijurica and @SanjnaSrivatsaVMware on this. We landed on having the spec be the place for ideal/compliant examples, but i absolutely think having a use case in the use case library that provides the workouts to start, and possibly adding this to a practitioner guide makes a ton of sense too. I'm going to take an AI to create the practitioner guide FR, so this point isn't lost, while fitting the examples to the ideal state.

Reviewer: @Matt-Cowsert 
I'm not totally following. 

> We landed on having the spec be the place for ideal/compliant examples

If that's the case, then why would we mention a situation for a data generator is potentially non-compliant, the way we do here?  Feels like it's inviting/expecting bad data generator behavior.

Reviewer 2: You're right — we shouldn't be inviting non-compliant behavior in the spec examples. I've struck the non-compliant row, the defensive OR pattern, and the two-approach comparison. The coverage example now shows three clean rows (uncovered, covered, ineligible) with a single straightforward calculation.

See commit b933a6f.

Author: That's an interesting thought. Unless we have a playbook in scope for this release, I'd recommend striking it out from here (or putting it in supporting content) so that we have it as a reference.

Acme Corp runs compute workloads on Aura Web and uses StackLens for observability monitoring. Some Aura Web compute usage is covered by a Resource Reservation. StackLens usage is uncovered but eligible for Interval Spend Commitments at monthly or annual terms. A practitioner wants to answer: "Which *commitment program* and provider should I target for my next purchase?"

Six usage rows for a single charge period (2025-04-01):

1. **Uncovered compute** (Rows 1-2): Two Aura Web Compute rows eligible for both FlexibleSpendPlan and ResourceReservation. [BilledCost](#datasets.costandusage.billedcost) totals &dollar;500.00 across both rows.
2. **Covered compute** (Row 3): Aura Web Compute covered by an existing ResourceReservation. Filtered out by the query because [CommitmentDiscountId](#datasets.costandusage.commitmentdiscountid) is populated.
3. **Ineligible support** (Row 4): Aura Web Support with no CommitmentProgramEligibilityDetails. Filtered out because the column is null.
4. **Uncovered observability** (Rows 5-6): Two StackLens Observability rows eligible for MonthlyIntervalSpendCommitment and AnnualIntervalSpendCommitment. BilledCost totals &dollar;200.00.

> The [eligible uncovered spend query](#supportedfeatures.commitmentprogrameligibilitydetails) expands the CommitmentPrograms array via CROSS JOIN UNNEST, then groups by provider, service, and ProgramType:

Reviewer: This doesn't feel like enough context.  This needs to presume that the reader is starting from scratch with stuff like this.  How about:

```suggestion
To evaluate these purchasing options, the `CommitmentPrograms` JSON array must be flattened so each eligible program can be analyzed independently. By expanding the array (e.g., via `CROSS JOIN UNNEST`) and grouping the uncovered costs by `ServiceProviderName`, `ServiceName`, and `EligibleProgramType`, the practitioner gets the summary presented in the table below. A reference implementation is available in the [eligible uncovered spend query](#supportedfeatures.commitmentprogrameligibilitydetails) supported feature.
```

> | ServiceProviderName | ServiceName | EligibleProgramType | TotalEligibleUncoveredCost |

Reviewer: Please drop the word "Total" from `TotalEligibleUncoveredCost` so it simply reads `EligibleUncoveredCost`.

A couple of reasons for this:
1. **Consistency with the spec:** The core specification avoids "Total" prefixes for financial metrics. We use `BilledCost` and `EffectiveCost`, not `TotalBilledCost`, so this keeps our naming conventions aligned.
2. **Redundancy:** Because the table represents data grouped by Provider, Service, and Program Type, the fact that the cost is an aggregate is already heavily implied by the context. "Total" is usually just a leftover artifact from writing the SQL query (`SUM(BilledCost) AS Total...`) and tends to clutter the presentation without adding new meaning.

```suggestion
| ServiceProviderName | ServiceName | EligibleProgramType | EligibleUncoveredCost |
```

Author: Accepting this
